### PR TITLE
[WIP] Codeinsights MVP backend

### PR DESCRIPTION
**Pull request by slimsag**
See #17218 for the general plan / work being done on code insights.

I am working from this `codeinsights` branch (which will be unstable/buggy/broken sometimes) and my aim is to have this branch by EOW run a very basic code insights backend in dev environments, including all the app infra/changes we need, running timescaledb, doing the DB migrations to create the schema, and the GraphQL schema/API/backend making requests and serving data from timescale.

This branch itself won't be merged, rather I will be taking changes from here once they are solid and sending individual PRs for review (so they meet our quality bar, are not "one big PR", etc.) I will be posting updates here as I send out those PRs.

TODO:

- [x] Grafana dashboards -- issue filed elsewhere
- [x] testing: An end-to-end test where we use the GraphQL API to modify user settings/add insights, and the GraphQL API to query data after a while? -- issue filed elsewhere
- [x] Investigate adding more tests for enterprise/internal/insights/background/* - see https://github.com/sourcegraph/sourcegraph/pull/18267 -- issue filed elsewhere
- [x] Basic webhook fetching support -- decided to adopt push model, see other issue
- [x] Deploy TimescaleDB in non-dev envs
- [ ] Generate schema.insights.md documentation
- [ ] `dev/db/squash_migrations.sh` starts Postgres 9.6 if the connected DB version != 9.6, so it doesn't work with these migrations.
- [ ] `internal/db/schemadoc/main.go` starts Postgres 9.6 if the connected DB version != 9.6, so it doesn't work with this DB schema yet.
- [ ] Add code TODO for retroactively updating repo names in DB

**Feb 21-26:**

- [x] insights: correct over-reporting / aggregation of data points #18632

**Feb 15-19:**

- [x] Basic search query execution & storage
- [x] Investigate using go-mockgen for GraphQL resolver tests: https://github.com/sourcegraph/sourcegraph/pull/18075#discussion_r572359571

**Feb 8-9:**

- [x] Implement DB store tests
- [x] Merge DB schema
- [x] Implement GraphQL backend tests

**Feb 1-5:**

- [x] Implement DB store
- [x] Implement GraphQL backend
- [x] Investigate replacing `repo_names` and `metadata` with https://docs.timescale.com/latest/using-timescaledb/compression (Tomas suggested, looks really cool) -- CONCLUSION: https://github.com/sourcegraph/sourcegraph/pull/17227#issuecomment-773704129
- [x] Start TImescaleDB in testing env; run migrations #17733 #17586

**Jan 18-22:**

- [x] Determine where to store insights (user/org/global settings? likely not DB for now, but file an issue for later?)
- [x] Run migrations on server startup
- [x] Plan DB schema & queries/inserts
- [x] Generate separate schema.md files for codeintel/frontend DBs so we can generate schema.insights.md later
- [x] Tag TimescaleDB Docker image as `sourcegraph/codeinsights-db`
- [x] Run `codeinsights-db` as part of dev server.
- [x] Add `migrations/codeinsights` DB migration foundation
- [x] Add GraphQL backend scaffolding / stubs

### CODE REVIEW

**Comment by slimsag** [comment]
PR to generate separate schema.md files for codeintel/frontend DBs so we can generate a schema.md for code insights DB schema soon: https://github.com/sourcegraph/sourcegraph/pull/17228

File: migrations/codeinsights/1000000001_initial_schema.up.sql [hunk 1]
```diff
@@ -0,0 +1,32 @@
+BEGIN;
+
+CREATE EXTENSION IF NOT EXISTS timescaledb;
+
+CREATE TABLE histogram_events (
```
Comment (felixfbecker) [inline migrations/codeinsights/1000000001_initial_schema.up.sql:5]: What does "histogram" mean in this context?

Comment (slimsag) [reply migrations/codeinsights/1000000001_initial_schema.up.sql:5]: Sorry, histogram was a poor name here, I have a commit which rewords this I haven't yet pushed up. What I actually mean here is effectively that this table stores events akin to a Prometheus Gauge (numbers that go up or down): https://prometheus.io/docs/concepts/metric_types/#gauge

I should note though, we benefit from Timescale's continuous aggregation described here: https://docs.timescale.com/latest/using-timescaledb/continuous-aggregates - so even though this table only stores events of some metric going up/down over time, for insights that display a single value we can either display the most recent value (with no change in API) or do more fancy aggregation (average/min/max of the value over time) efficiently if we want to in the future (by exposing another GraphQL API to access that info)

File: migrations/codeinsights/1000000001_initial_schema.up.sql [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    -- The timestamp of the recorded event.
+    time TIMESTAMPTZ NOT NULL,
+
+    -- The floating point value at the time of the event.
+    value double precision NOT NULL,
+
+    -- Metadata about this event, this can be any arbitrary JSON metadata which will be returned
+    -- when querying events, but cannot be filtered on.
+    metadata jsonb NOT NULL,
```
Comment (slimsag) [inline migrations/codeinsights/1000000001_initial_schema.up.sql:14]: Ongoing discussion about this: https://sourcegraph.slack.com/archives/C014ZCKMCAV/p1610764544015200

File: migrations/codeinsights/1000000001_initial_schema.up.sql [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    -- If the event was for a single repository (usually the case) then this field should indicate
+    -- the repository ID at the time the event was created. Note that the repository may no longer
+    -- exist / be valid at query time, however.
+    repo_id integer,
+
+    -- If the event was for a single repository (usually the case) then this field should indicate
+    -- the repository name at the time the event was created. Note that the repository name may
+    -- have changed since the event was created (e.g. if the repo was renamed), in which case this
+    -- describes the outdated repository na,e.
+    repo_name citext
```
Comment (slimsag) [inline migrations/codeinsights/1000000001_initial_schema.up.sql:25]: Ongoing discussions about this:

* repo vs. global vs. other association of events: https://sourcegraph.slack.com/archives/C014ZCKMCAV/p1610763511014200
* what happens if repositories get renamed/removed/etc: https://sourcegraph.slack.com/archives/C014ZCKMCAV/p1610762810012800

**Comment by slimsag** [comment]
PR to tag the TimescaleDB image as `sourcegraph/codeinsights-db`: https://github.com/sourcegraph/sourcegraph/pull/17427

**Comment by slimsag** [comment]
PR to run codeinsights-db (TimescaleDB) as part of dev server: https://github.com/sourcegraph/sourcegraph/pull/17431

**Comment by slimsag** [comment]
PR to add codeinsights-db migrations foundation: https://github.com/sourcegraph/sourcegraph/pull/17432

**Comment by slimsag** [comment]
Update: There is some more interesting stuff in here now, in particular look at `migrations/codeinsights/1000000001_initial_schema.up.sql` and `README.codeinsights.md` for some major changes to the DB schema

**Comment by slimsag** [comment]
PR to add GraphQL backend scaffolding: https://github.com/sourcegraph/sourcegraph/pull/17497

File: migrations/codeinsights/1000000001_initial_schema.up.sql [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+CREATE INDEX metadata_metadata_gin ON metadata USING GIN (metadata);
+
+-- Records events over time associated with a repository (or none, i.e. globally) where a single
+-- numerical value is going arbitrarily up and down.
+--
+-- Repository association is based on both repository ID and name. The ID can be used to refer to
+-- a specific repository, or lookup the current name of a repository after it has been e.g. renamed.
+-- The name can be used to refer to the name of the repository at the time of the event's creation,
+-- for example to trace the change in a gauge back to a repository being renamed.
+CREATE TABLE gauge_events (
```
Comment (slimsag) [inline migrations/codeinsights/1000000001_initial_schema.up.sql:57]: insights_series

**Comment by slimsag** [comment]
PR to avoid running TimescaleDB migrations against singleton postgres dev/testing instance (which would fail): #17585

**Comment by slimsag** [comment]
PR: dev: correct codeinsights-db port; add connection info #17588

**Comment by slimsag** [comment]
PR: execute database migrations #17586

**Comment by slimsag** [comment]
PR to fix a nasty dev issue we've had for a while (turns out we weren't testing the right sourcegraph version with e2e tests?) which caused me to revert and re-land some changes: https://github.com/sourcegraph/sourcegraph/pull/17731

**Comment by slimsag** [comment]
PR to add basic store package + testing infrastructure https://github.com/sourcegraph/sourcegraph/pull/17733

**Comment by slimsag** [comment]
PR to prevent execution in sourcegraph/server: https://github.com/sourcegraph/sourcegraph/pull/17735

File: migrations/codeinsights/1000000001_initial_schema.up.sql (hunk 4 above)
Comment (slimsag) [reply migrations/codeinsights/1000000001_initial_schema.up.sql]: Renamed to series_points as I think that is sightly better (given we are in the `codeinsights-db` DB, reiterating `insights` seems silly)

**Comment by slimsag** [comment]
PR to add new GraphQL schema: https://github.com/sourcegraph/sourcegraph/pull/17842

**Comment by slimsag** [comment]
PR to improve backend testing experience: https://github.com/sourcegraph/sourcegraph/pull/17843

**Comment by slimsag** [comment]
PR for new user/org/global settings schema definition: https://github.com/sourcegraph/sourcegraph/pull/17844

**Comment by slimsag** [comment]
## Assessment of whether or not using TimescaleDB's [compression feature](https://docs.timescale.com/latest/using-timescaledb/compression) makes sense.

Challenge/question: Does it make sense to use TimescaleDB's compression feature? The thinking here is that it would let us in theory merely insert directly the repository names and metadata, instead of inserting that data to another table for deduplication and inserting just IDs to our time-series table as shown [here](https://github.com/sourcegraph/sourcegraph/blob/145c688a21e98701cfe513e60eb6758324a62007/migrations/codeinsights/1000000001_initial_schema.up.sql#L7-L42).

I spent ~1h looking into this.

Notable learnings:

* We would need to choose a `segmentby` option for the compression (see [docs here](https://docs.timescale.com/latest/using-timescaledb/compression#understanding-segmentby)). They describe how they use a btree index on the segmentby column so that lookup is fast on it, but other columns will be compressed effectively into an array and require decompressing in order to filter on with a `WHERE` cause, making it quite slow. "In particular, queries that reference the segmentby columns in the WHERE clause are very efficient. Thus, it is important to pick the correct set of segmentby columns. "
* I think in our case we are banking on the idea that we can filter by at least repository and metadata, so those would need to be our segmentby columns. Metadata has another problem (described below), but I think this means in practice that the only real thing we would be compressing is the recorded `value` - most other fields would not be compressed.
* That we store `repo_name` ("latest known name for the repo") and `original_repo_name` ("repo name known at time of insert") would mean `original_repo_name` could be compressed.
* "Most indexes set on the hypertable are removed/ignored when reading from compressed chunks! TimescaleDB creates and uses custom indexes to incorporate the segmentby and orderby parameters during compression." I believe means we would not be able to use a GIN index on the jsonb metadata column. Potentially we could replace this column with non-jsonb ones, but really we need to prove out the product itself a bit more before we will know.
* It would be very annoying to go back over the data set and update repository names to account for repo names, as they make this remark in their docs: "One of the current limitations of TimescaleDB is that once chunks are converted into compressed column form, we do not currently allow any further modifications of the data (e.g., inserts, updates, deletes) or the schema without manual decompression. In other words, chunks are immutable in compressed form. "
* This remark seems to indicate we would encounter annoyances/difficulties upgrading our schema in the future if using compression: "WARNING: The current release of TimescaleDB supports the ability to query data in compressed chunks. However, it does not support inserts or updates into compressed chunks. We also block modifying the schema of hypertables with compressed chunks."
* I think the primary concern with schema changes would be "do customers have enough disk space to run the migration, which decompresses everything, alters the table schema, then recompresses?"
* We would need to select a time after which to compress data, and in their docs they remark that this has implications on how you can query your data. In specific, they talk about using a 7-day timeframe to allow for querying e.g. CPU usage of specific servers, and then (approximately) only querying non-server-specific information after that 7d timeframe. With insights, we need to query specific information ("give me data for just this one repo, over all past time") which seems incompatible with this.

## Conclusion

I think TimescaleDB compression would lock us into some hard choices, and further tie us to Timescale which may not be a good thing if we later find vanilla Postgres is something we need to switch to. It's compression feature seems more tailored towards truly large amounts of low-cardinality time series data - and we have the opposite problem (few data points, large cardinality.)

It seems easy to adopt compression, but not necessarily easy to unadopt it - so for now I have chosen to continue on the path of using a more traditional SQL schema and skip their compression feature.

**Comment by slimsag** [comment]
PR to prepare for having DB/integration tests for GraphQL resolvers: https://github.com/sourcegraph/sourcegraph/pull/18001

**Comment by slimsag** [comment]
PR to implement GraphQL resolvers: #18075

**Comment by slimsag** [comment]
dev: make drop-entire-db script also drop insights DB #18081

**Comment by slimsag** [comment]
PR: add & prepare to use generated store mocks #18118

**Comment by slimsag** [comment]
* PR: insights: add very basic GraphQL resolver integration test #18035
* PR: insights: add initial DB schema #18083

**Comment by slimsag** [comment]
* PR: insights: store: add support for querying series data points #18129
* PR: insights: resolvers: implement point querying & add integration tests #18130

**Comment by slimsag** [comment]
* insights: change series_id to string (hash) #18230

**Comment by slimsag** [comment]
* insights: store: query metadata & other minor improvements #18254

**Comment by slimsag** [comment]
* insights: store: add support for recording data points #18255

**Comment by slimsag** [comment]
insights: add new discovery package for locating insights (and use it) #18264

**Comment by slimsag** [comment]
* insights: store: query metadata & other minor improvements #18254

**Comment by slimsag** [comment]
* migrations: add insights_query_runner_jobs table #18265

**Comment by slimsag** [comment]
# Data size & retention policy speculation

In [the original RFC](https://docs.google.com/document/d/1LX6Msbxnz_DRf6OItZLaX-vELfnHGKgODkb1bVauolk/edit#heading=h.q9xc4jvjs2ld) we speculated the backend:

* Must support insights over 16,000 repos.
* Ideally over our largest customers ~500k+ repos.
* We speculated customers would at max have ~500 dashboards (spread across ~2k or so devs)
* We speculated there may be 5 panels per dashboard, each with a one to a handful of data series recorded per-repo.
* Examples of series cardinality included:
    * A panel showing 20 java versions, recorded per repo
    * A panel showing 40 languages, recorded per repo 
    * A panel showing PR states(open/close/merged), recorded per repo - assuming a max of like 20,000 PRs.
* In an ideal world, keeping data for some long period of time like 6-12mo (but maybe with limitations as we start to talk about older data.)

We will obviously need horizontal scaling in the future to achieve the much larger scales (which should be easy to add with TimescaleDB, but I haven't looked into it.)

In theory, the best interval at which to record data points would be whenever repositories receive new commits. We are doing global searches (currently) at some interval, though, as that is a fair bit easier to implement - and so a question is at what interval we should do that for all repositories.

I would speculate that given all repositories on a Sourcegraph instance - up to 500k - we are unlikely to see a commit push frequency higher than every 5 minutes in general.

We can then speculate on the # of table rows per recording interval based on:

* 20-500k repos
* 10-500 dashboards with ~5 panels each (50-2500 panels)
* 5-40 semi-unique series per dashboard (let's say 50% are reused / same queries.)
* # of table rows per recording interval == repos * dashboard_panels * unique_series_per_dashboard

Some examples:

* 20k_repos*50_panels*5_series == 20,000,000 rows / recording interval (e.g. 5 minutes)
* 500k_repos*2500_panels*40_series == 50,000,000 rows/recording interval (e.g. 5 minutes)

Obviously that's a _lot_ of data and will most likely not "just work" magically. https://docs.timescale.com/latest/faq#scaling says they regularly test with 10+ billion rows and inserting 100-200k rows / second. I we assume we have a budget of 10 billion rows, and recorded 20mil/5min we would only be able to store the last 500 recording intervals (1.7 days).

But that's all very speculative, let's try some real

## Quick benchmarks

Inserting 1 year of data points for a _single repository, with no metadata_ at a 5 minute interval takes 6-10s using:

```
INSERT INTO series_points(
    time,
    series_id,
    value,
    metadata_id,
    repo_id,
    repo_name_id,
    original_repo_name_id)
SELECT time,
    0,
    random()*80 - 40,
    (SELECT id FROM metadata WHERE metadata = '{"hello": "world", "languages": ["Go", "Python", "Java"]}'),
    2,
    (SELECT id FROM repo_names WHERE name = 'github.com/gorilla/mux-renamed'),
    (SELECT id FROM repo_names WHERE name = 'github.com/gorilla/mux-original')
    FROM generate_series(TIMESTAMP '2021-01-01 00:00:00', TIMESTAMP '2022-01-01 00:00:00', INTERVAL '5 min') AS time;
```

Idle memory / CPU usage is quite minimal (41 MiB / 1% CPU), still, and disk usage grows by about 60 MiB. Most of the costs outside of disk storage will probably be incurred at query time - which can be helped via precomputed aggregations Timescale offers.

But the 60 MiB disk usage gives us a pretty good idea: 20,000 repos * 60 MiB == ~1171 GiB of disk storage for 1 year of data points recorded every 5min for 20k repos.

Based on this, I will go with a default recording interval of 10 minutes (we can adjust this if needed later) which would give us ~40k repos worth of data points for a single series type on a reasonable deployment, i.e. it gets us decently far for the initial version.

The next important things to look at for scaling are:

* A data retention policy that deletes data points after 1yr.
* Continuous aggregation, we may be able to do this and keep lesser precision historical data but still at a per-repo level (e.g. one day point per day if older than 3mo.)
* Keep repository-level metrics for the past ~3 months, but then only keep global (non-repo-specific) insights after that.

Lots of stuff/options to explore here in the TimescaleDB docs.

**Comment by slimsag** [comment]
insights: actually enqueue/record insights from global settings #18314

**Comment by slimsag** [comment]
* ci: build docker-images/codeinsights-db #18350
* dev: allow codeinsights-db more time to startup #18359
* insights: turn on for Sourcegraph.com #18360

**Comment by slimsag** [comment]
insights: correct over-reporting / aggregation of data points #18632

**Pull request closed without merging**